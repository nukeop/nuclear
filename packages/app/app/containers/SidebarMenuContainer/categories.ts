import { SemanticICONS } from 'semantic-ui-react/dist/commonjs/generic';

type MenuItem = {
    name: string;
    path: string;
    icon: SemanticICONS;
}

type Category = {
    name: string;
    items: MenuItem[];
}

export default [
  {
    name: 'main',
    items: [
      { name: 'dashboard', path: 'dashboard', icon: 'dashboard' },
      { name: 'downloads', path: 'downloads', icon: 'download' },
      { name: 'lyrics', path: 'lyrics', icon: 'microphone' },
      { name: 'plugins', path: 'plugins', icon: 'flask' },
      { name: 'search', path: 'search', icon: 'search' },
      { name: 'settings', path: 'settings', icon: 'cog' },
      { name: 'equalizer', path: 'equalizer', icon: 'align right' },
      { name: 'visualizer', path: 'visualizer', icon: 'tint' }
    ]
  }, {
    name: 'collection',
    items: [
      { name: 'favorite-albums', path: 'favorites/albums', icon: 'dot circle' },
      { name: 'favorite-tracks', path: 'favorites/tracks', icon: 'music' },
      { name: 'favorite-artists', path: 'favorites/artists', icon: 'user' },
      { name: 'library', path: 'library', icon: 'file audio outline' },
      { name: 'playlists', path: 'playlists', icon: 'list alternate outline' }
    ]
  }
] as Category[];
