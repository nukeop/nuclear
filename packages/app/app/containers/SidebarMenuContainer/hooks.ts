import { useCallback, useState } from 'react';

type UseResizablePanelArgs = {
  initialWidth: number;
  collapsedWidth: number;
  collapseThreshold: number;
  isCollapsed: boolean;
  onCollapse: (isCollapsed: boolean) => void;
};

export const useResizablePanel = ({
  initialWidth,
  collapsedWidth,
  collapseThreshold,
  isCollapsed,
  onCollapse
}: UseResizablePanelArgs) => {
  const [width, setWidth] = useState(initialWidth);

  const onSetWidth = useCallback(
    (width) => {
      if (!isCollapsed && width < collapseThreshold) {
        onCollapse(true);
        setWidth(collapsedWidth);
      }

      if (isCollapsed && width >= collapseThreshold) {
        onCollapse(false);
        setWidth(collapseThreshold);
      } else {
        setWidth(width);
      }
    },
    [collapseThreshold, collapsedWidth, isCollapsed, onCollapse]
  );

  return {
    width,
    onSetWidth
  };
};
