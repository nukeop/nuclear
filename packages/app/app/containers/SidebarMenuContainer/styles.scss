@import "../../vars.scss";

.sidebar_menu_container {
    z-index: 11;

    display: flex;
    flex: 0 0 auto;
    flex-flow: column;

    background-color: $background;
    box-shadow: 5px 0px 5px 0px rgba(0, 0, 0, 0.18), 5px 0px 20px 0px rgba(0, 0, 0, 0.09);

    &.collapsed {
        width: 3rem;

        .sidebar_brand {
            padding: 1rem 0;
        }

        .sidebar_menu_item_container {
            display: flex;
            justify-content: center;
            padding: 0.5rem 0.25rem;

            i {
                margin: 0;
            }
        }

        .active_nav_link {
            border: none;
            background: rgba($background3, 0.6);
        }

        span {
            margin: 0 !important;
        }
    }

    .content {
        overflow: hidden;
    }
}

.active_nav_link {
    color: $white !important;
    border-left: 0.5rem solid $green;
}
