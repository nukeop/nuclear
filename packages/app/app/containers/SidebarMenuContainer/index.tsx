import React, { useCallback } from 'react';
import { NavLink } from 'react-router-dom';
import { useTranslation } from 'react-i18next';
import { useDispatch, useSelector } from 'react-redux';
import { Icon } from 'semantic-ui-react';
import cx from 'classnames';
import { ResizablePanel } from '@nuclear/ui';

import SidebarMenu from '../../components/SidebarMenu';
import SidebarMenuCategoryHeader from '../../components/SidebarMenu/SidebarMenuCategoryHeader';
import SidebarMenuItem from '../../components/SidebarMenu/SidebarMenuItem';
import { settingsSelector } from '../../selectors/settings';
import { setBooleanOption } from '../../actions/settings';
import { useResizablePanel } from './hooks';
import categories from './categories';
import styles from './styles.scss';
// import { UserPanelContainer } from '../UserPanelContainer';

const SidebarMenuContainer: React.FC = () => {
  const { t } = useTranslation('app');
  const dispatch = useDispatch();
  const settings = useSelector(settingsSelector);
  const isCollapsed = settings.compactMenuBar;

  const onCollapse = useCallback((collapsedState: boolean) => {
    dispatch(setBooleanOption('compactMenuBar', collapsedState));
  }, [dispatch]);
  const {
    width,
    onSetWidth
  } = useResizablePanel({
    initialWidth: 300,
    collapsedWidth: 42,
    collapseThreshold: 200,
    onCollapse,
    isCollapsed
  });

  return <ResizablePanel
    width={width}
    onSetWidth={onSetWidth}
    isCollapsed={isCollapsed}
    collapsedWidth={42}
    classes={
      {
        root: cx(
          styles.sidebar_menu_container,
          { [styles.collapsed]: isCollapsed }
        ),
        content: styles.content
      }
    }
  >
    <SidebarMenu>
      {
        categories.map(({ name, items }) => (
          <React.Fragment
            key={name}
          >
            <SidebarMenuCategoryHeader
              compact={settings.compactMenuBar}
              headerText={t(name)}
            />
            {
              items.map(({ name, path, icon }) => (
                <NavLink key={path} to={'/' + path} activeClassName={styles.active_nav_link}>
                  <SidebarMenuItem
                    name={t(name)}
                    compact={settings.compactMenuBar}
                  >
                    <Icon name={icon} />{!settings.compactMenuBar && t(name)}
                  </SidebarMenuItem>
                </NavLink>
              ))}
          </React.Fragment>
        ))}
    </SidebarMenu>
    {/* {
      !isCollapsed &&
      <UserPanelContainer />
    } */}
  </ResizablePanel>;
};

export default SidebarMenuContainer;
