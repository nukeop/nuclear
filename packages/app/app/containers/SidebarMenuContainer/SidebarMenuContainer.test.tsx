import React from 'react';

import SidebarMenuContainer from '.';
import { buildStoreState } from '../../../test/storeBuilders';
import { setupI18Next, mountComponent } from '../../../test/testUtils';

describe('Sidebar menu container', () => {
  beforeAll(() => {
    setupI18Next();
  });

  it('should render the sidebar menu', () => {
    const { component } = mountApp();
    const sidebar = component.getByTestId('sidebar-menu');

    expect(sidebar).toMatchSnapshot();
  });

  it.each([
    ['dashboard', /dashboard/i, 'dashboard'],
    ['downloads', /downloads/i, 'downloads'],
    ['favorite albums', /favorite albums/i, 'favorites/albums'],
    ['favorite tracks', /favorite tracks/i, 'favorites/tracks'],
    ['playlists', /playlists/i, 'playlists'],
    ['plugins', /plugins/i, 'plugins'],
    ['settings', /settings/i, 'settings'],
    ['search', /search/i, 'search'],
    ['lyrics', /lyrics/i, 'lyrics'],
    ['equalizer', /equalizer/i, 'equalizer'],
    ['visualizer', /visualizer/i, 'visualizer'],
    ['local library', /local library/i, 'library']
  ])('should go to %s on click', (text, regex, path) => {
    const { component, history } = mountApp();

    const visualizer = component.getByText(regex);
    visualizer.click();

    expect(history.location.pathname).toBe(`/${path}`);
  });

  const mountApp = () => mountComponent(
    <SidebarMenuContainer />,
    ['/'],
    buildStoreState()
      .build()
  );
});
