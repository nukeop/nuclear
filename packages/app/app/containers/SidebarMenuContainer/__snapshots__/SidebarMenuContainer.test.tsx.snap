// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Sidebar menu container should render the sidebar menu 1`] = `
<div
  class="sidebar_menu"
  data-testid="sidebar-menu"
>
  <div
    class="sidebar_menu_category_header"
  >
    <span>
      Main
    </span>
    <hr />
  </div>
  <a
    href="/dashboard"
  >
    <div
      class="sidebar_menu_item_container"
    >
      <i
        aria-hidden="true"
        class="dashboard icon"
      />
      Dashboard
    </div>
  </a>
  <a
    href="/downloads"
  >
    <div
      class="sidebar_menu_item_container"
    >
      <i
        aria-hidden="true"
        class="download icon"
      />
      Downloads
    </div>
  </a>
  <a
    href="/lyrics"
  >
    <div
      class="sidebar_menu_item_container"
    >
      <i
        aria-hidden="true"
        class="microphone icon"
      />
      Lyrics
    </div>
  </a>
  <a
    href="/plugins"
  >
    <div
      class="sidebar_menu_item_container"
    >
      <i
        aria-hidden="true"
        class="flask icon"
      />
      Plugins
    </div>
  </a>
  <a
    href="/search"
  >
    <div
      class="sidebar_menu_item_container"
    >
      <i
        aria-hidden="true"
        class="search icon"
      />
      Search Results
    </div>
  </a>
  <a
    href="/settings"
  >
    <div
      class="sidebar_menu_item_container"
    >
      <i
        aria-hidden="true"
        class="cog icon"
      />
      Settings
    </div>
  </a>
  <a
    href="/equalizer"
  >
    <div
      class="sidebar_menu_item_container"
    >
      <i
        aria-hidden="true"
        class="align right icon"
      />
      Equalizer
    </div>
  </a>
  <a
    href="/visualizer"
  >
    <div
      class="sidebar_menu_item_container"
    >
      <i
        aria-hidden="true"
        class="tint icon"
      />
      Visualizer
    </div>
  </a>
  <div
    class="sidebar_menu_category_header"
  >
    <span>
      Collection
    </span>
    <hr />
  </div>
  <a
    href="/favorites/albums"
  >
    <div
      class="sidebar_menu_item_container"
    >
      <i
        aria-hidden="true"
        class="dot circle icon"
      />
      Favorite Albums
    </div>
  </a>
  <a
    href="/favorites/tracks"
  >
    <div
      class="sidebar_menu_item_container"
    >
      <i
        aria-hidden="true"
        class="music icon"
      />
      Favorite Tracks
    </div>
  </a>
  <a
    href="/favorites/artists"
  >
    <div
      class="sidebar_menu_item_container"
    >
      <i
        aria-hidden="true"
        class="user icon"
      />
      Favorite Artists
    </div>
  </a>
  <a
    href="/library"
  >
    <div
      class="sidebar_menu_item_container"
    >
      <i
        aria-hidden="true"
        class="file audio outline icon"
      />
      Local Library
    </div>
  </a>
  <a
    href="/playlists"
  >
    <div
      class="sidebar_menu_item_container"
    >
      <i
        aria-hidden="true"
        class="list alternate outline icon"
      />
      Playlists
    </div>
  </a>
</div>
`;
